### PR TITLE
feat: add support for cohere transcribe

### PR DESCRIPTION
**Pull request by sammcj**
Adds a new `cohere` feature flag with a [Cohere Transcribe](https://docs.cohere.com/docs/transcribe) ASR engine - _"an open source research release of a 2B parameters dedicated audio-in, text-out, automatic speech recognition (ASR) model. The model supports a total of 14 languages."_

- Takes raw 16 kHz mono f32 audio and returns transcribed text
- Supports 14 languages via explicit ISO 639-1 language selection
- Configurable punctuation and inverse text normalisation (ITN)
- Configurable max decode tokens (default 512, capped at the model's 1024 limit)
- No timestamps - the model doesn't support them

## Files

- `src/cohere.rs` - public `CohereASR` struct with `from_pretrained()` and `transcribe_audio()` methods
- `src/model_cohere.rs` - ONNX session wrappers for encoder and merged decoder
- `src/lib.rs` - feature-gated module declarations and re-export
- `Cargo.toml` - `cohere` feature flag

## Usage

```rust
use parakeet_rs::{CohereASR, ExecutionConfig};

let mut asr = CohereASR::from_pretrained("./cohere-onnx", None)?;
let text = asr.transcribe_audio(&audio_samples, "en", true, true)?;
```

## Design

Follows the same conventions as the existing Sortformer, Multitalker, and Nemotron features:

- Feature-gated with `#[cfg(feature = "cohere")]` throughout
- Internal model types use `pub(crate)` visibility (only `CohereASR` is exported)
- Mel features computed in Rust via the shared `src/audio.rs` utilities (`extract_features_raw`) using the model's `preprocessor_config.json` parameters
- Tokenisation via the `tokenizers` crate loading `tokenizer.json` directly - no custom vocab parser
- Greedy decoding with n-gram repetition detection to bail out of pathological loops
- Zero-copy `TensorRef` views for the encoder output and KV cache tensors in the decoder loop
- Consumes the HuggingFace-standard ONNX export produced by Optimum: `input_features` in, `last_hidden_state` out, merged decoder with growing `past_key_values.N.{decoder,encoder}.{key,value}` cache, `position_ids`, and `attention_mask`
- Canonical 10-token decoder prompt matching what `CohereAsrProcessor` produces: `[▁, <|startofcontext|>, <|startoftranscript|>, <|emo:undefined|>, <|src_lang|>, <|tgt_lang|>, <|pnc|>/<|nopnc|>, <|itn|>/<|noitn|>, <|notimestamp|>, <|nodiarize|>]`

## Model files

The model directory must contain (flat or under `onnx/`):

- `encoder_model[_quantized|_fp16|_q4|_q4f16].onnx` (+ `.onnx_data` companions for external weights)
- `decoder_model_merged[_quantized|_fp16|_q4|_q4f16].onnx` (+ `.onnx_data`)
- `tokenizer.json`
- `preprocessor_config.json`

This layout matches [`onnx-community/cohere-transcribe-03-2026-ONNX`](https://huggingface.co/onnx-community/cohere-transcribe-03-2026-ONNX), which ships FP32, FP16, INT8, INT4, and INT4/FP16 variants of the model.

To produce your own export from the upstream PyTorch checkpoint, install [`optimum`](https://github.com/huggingface/optimum) and run:

```sh
optimum-cli export onnx \
    --model CohereLabs/cohere-transcribe-03-2026 \
    --task automatic-speech-recognition-with-past \
    ./cohere-onnx
```

No custom export script is needed - the `cohere_asr` model type is supported by Optimum's standard exporter, so this slots into the HuggingFace ecosystem without any bespoke conversion code.

## Test plan

- [x] `RUSTFLAGS="-D warnings" cargo clippy --features cohere` passes clean
- [x] `cargo test --features cohere` - 17 tests pass (argmax, supported languages count, n-gram repetition, plus the existing crate tests)
- [x] `cargo clippy` / `cargo test` without the cohere feature - no regressions (14 tests pass)
- [x] `cargo check --features sortformer` / `--features multitalker` - no regressions
- [x] End-to-end transcription against `onnx-community/cohere-transcribe-03-2026-ONNX` INT8 - output matches the HuggingFace reference `model.generate()` on the same audio, confirming the prompt format, mel extraction, decoder cache management, and detokenisation are all correct

### CODE REVIEW

File: src/cohere.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+            // Format: "token_text id"
+            let Some((token, id_str)) = line.rsplit_once(' ') else {
+                continue;
+            };
+            let Ok(id) = id_str.parse::<usize>() else {
+                continue;
+            };
+
+            // Extend vocab vector if needed
+            while vocab.len() <= id {
```
Comment (altunenes) [inline src/cohere.rs]: supported_languages() returned 199 codes because this parser treats every <|xx|> length-6 token as a supported lang it seems wghile tokens.txt defininig placeholders for the full range when I chevk you HF page it say cohere only ships trained weights for 14 languages per the model card.   I applied a small allowlist patch locally to confirm:

const SUPPORTED_LANGUAGES: &[&str] = &[  
 "en", "fr", "de", " //bla bla all langs...

```rust
if token.starts_with("<|") && token.ends_with("|>") && token.len() == 6 {    
let lang_code = &token[2..4];
if SUPPORTED_LANGUAGES.contains(&lang_code) { 
lang_tokens.insert(lang_code.to_string(), id as i64); 

```
After this change supported_languages() returns exactly the 14 documented codes... Could you fold this in?

**Review by altunenes — APPROVED** [review]
really fast model thank you. maybe a simple change regarding with the supported_languages() ?  

and also CoreML is still problematic with this model lol, not surprising :-P 

also is it possible to include onnx script? and also maybe base models on your hf also (not only int8 ones)?

File: src/cohere.rs (hunk 1 above)
Comment (sammcj) [reply src/cohere.rs]: Good call, added now.

**Comment by sammcj** [comment]
> also is it possible to include onnx script? and also maybe base models on your hf also (not only int8 ones)?

The INT8 repo on HF is just a copy of the weights cjpais's quantised and uses for [Handy](https://handy.computer/), I'll ask if he has a script he can share.

**Comment by altunenes** [comment]
I see. lets wait a lil bit if no response we can merge it and add the script later on maybe... will wait for a response from you.

Handy btw looks really cool

**Comment by sammcj** [comment]
Spoke to cjpais, turns out he actually copied his as well (Perhaps from [onnx-community/cohere-transcribe-03-2026-ONNX](https://huggingface.co/onnx-community/cohere-transcribe-03-2026-ONNX/tree/main) which for whatever reason I didn't realise existed, they publish the fp16, q4fp16 and q4 weights).

I think the need for a conversion script is now mitigated, however out of curiosity I've asked [on their HF repo](https://huggingface.co/onnx-community/cohere-transcribe-03-2026-ONNX/discussions/1).

*Edit: I just looked into the onnx-community weights and they're actually quite different so I suspect he may have used another source

---

Re: Handy - yes it really is brilliant software, he's done a fantastic job of making a STT tool that's lightweight, fast, easy to use and flexible - I use it all throughout the day. He's also a friendly and helpful engineer, although Handy isn't using parakeet-rs I've chatted with him a few times about it when I was working on the PRs here in the past.

**Comment by sammcj** [comment]
I looked into using onnx-community/cohere-transcribe-03-2026-ONNX, this export uses a fundamentally different ONNX interface:

  |                | This PR (_currently_)               | onnx-community export                                       |
  |----------------|-----------------------------------|-------------------------------------------------------------|
  | Encoder input  | raw 16kHz PCM                     | pre-computed 128-bin log-mel                                |
  | Encoder output | per-layer cross K/V pre-projected | last_hidden_state                                           |
  | Decoder cache  | fixed buffer + scalar offset      | growing past_key_values.N.X.X + position_ids/attention_mask |

I'm going to mark this as draft while I assess if it's worth changing our cohere implementation to align with the onnx-community export (and check what other recent exports have been released).

**Comment by sammcj** [comment]
@altunenes after some investigation I've refactored the code to be better aligned with the onnx-community exports as I believe that to be a more standard approach.

I do not believe an export script is needed as this now supports the HF-standard export format, and can produced by a single command (comment added to the code for this):

```
optimum-cli export onnx \
  --model CohereLabs/cohere-transcribe-03-2026 \
  --task automatic-speech-recognition-with-past \
  ./out
```

**Review by altunenes — APPROVED** [review]
thanks for pushing this through a second iteration :-)
